File: src/main/java/web/controlevacinacao/model/Usuario.java
package web.controlevacinacao.model;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.SequenceGenerator;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import web.controlevacinacao.service.NomeUsuarioUnicoService;
import web.controlevacinacao.validation.UniqueValueAttribute;

@Entity
@Table(name = "usuario")
@UniqueValueAttribute(attribute = "nomeUsuario", service = NomeUsuarioUnicoService.class, message = "Já existe um nome de usuário igual a este cadastrado")
public class Usuario implements Serializable {

	private static final long serialVersionUID = 1L;

	@Id
	@SequenceGenerator(name="gerador55", sequenceName="usuario_codigo_seq", allocationSize=1)
	@GeneratedValue(generator="gerador55", strategy=GenerationType.SEQUENCE)
	private Long codigo;
	@NotBlank(message = "O nome do usuário é obrigatório")
	private String nome;
	@NotBlank(message = "O e-mail do usuário é obrigatório")
	private String email;
	@NotBlank(message = "A senha do usuário é obrigatória")
	private String senha;
	@Column(name = "nome_usuario")
	@NotBlank(message = "O nome de usuário do usuário é obrigatório")
	private String nomeUsuario;
	@Column(name = "data_nascimento")
	@NotNull(message = "A data de nascimento do usuário é obrigatória")
	private LocalDate dataNascimento;
	private boolean ativo;
	@ManyToMany
	@JoinTable(name = "usuario_papel", joinColumns = @JoinColumn(name = "codigo_usuario"), inverseJoinColumns = @JoinColumn(name = "codigo_papel"))
	@Size(min = 1, message = "O usuário deve ter ao menos um papel no sistema")
	private List<Papel> papeis = new ArrayList<>();

	public Long getCodigo() {
		return codigo;
	}

	public void setCodigo(Long codigo) {
		this.codigo = codigo;
	}

	public String getNome() {
		return nome;
	}

	public void setNome(String nome) {
		this.nome = nome;
	}

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getSenha() {
		return senha;
	}

	public void setSenha(String senha) {
		this.senha = senha;
	}

	public String getNomeUsuario() {
		return nomeUsuario;
	}

	public void setNomeUsuario(String nomeUsuario) {
		this.nomeUsuario = nomeUsuario;
	}

	public LocalDate getDataNascimento() {
		return dataNascimento;
	}

	public void setDataNascimento(LocalDate dataNascimento) {
		this.dataNascimento = dataNascimento;
	}

	public boolean isAtivo() {
		return ativo;
	}

	public void setAtivo(boolean ativo) {
		this.ativo = ativo;
	}

	public void adicionarPapel(Papel papel) {
		papeis.add(papel);
	}

	public void removerPapel(Papel papel) {
		papeis.remove(papel);
	}

	public List<Papel> getPapeis() {
		return papeis;
	}

	public void setPapeis(List<Papel> papeis) {
		this.papeis = papeis;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((codigo == null) ? 0 : codigo.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Usuario other = (Usuario) obj;
		if (codigo == null) {
			if (other.codigo != null)
				return false;
		} else if (!codigo.equals(other.codigo))
			return false;
		return true;
	}

	@Override
	public String toString() {
		return "codigo: " + codigo + "\nnome: " + nome + "\nemail: " + email + "\nsenha: " + senha + "\nusuario: " + nomeUsuario + "\ndataNascimento: " + dataNascimento + "\nativo: " + ativo;
	}

}

File: src/main/kotlin/web/fidex/model/fidex_model/Usuario.kt
package web.fidex.model.fidex_model

import jakarta.persistence.*
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Size
import java.io.Serializable
import java.time.LocalDate
import java.util.*

@Entity
@Table(name = "usuario")
class Usuario(
        @Id
        @SequenceGenerator(
                name = "gerador55",
                sequenceName = "usuario_codigo_seq",
                allocationSize = 1
        )
        @GeneratedValue(generator = "gerador55", strategy = GenerationType.SEQUENCE)
        var codigo: Long? = null,
        @field:NotBlank(message = "O nome é obrigatório") var nome: String? = null,
        @field:NotBlank(message = "O e-mail é obrigatório")
        @field:Email(message = "O e-mail é inválido")
        var email: String? = null,
        @field:NotBlank(message = "A senha é obrigatória")
        @field:Size(min = 8, message = "A senha deve ter no mínimo 8 caracteres")
        var senha: String? = null,
        @Column(name = "nome_usuario")
        @field:NotBlank(message = "O nome de usuário é obrigatório")
        @field:Size(min = 4, message = "O usuário deve ter no mínimo 4 caracteres")
        var nomeUsuario: String? = null,
        @Column(name = "data_nascimento") var dataNascimento: LocalDate? = null,
        var ativo: Boolean = false,
        var cashback: Double? = null,
        @ManyToMany
        @JoinTable(
                name = "usuario_papel",
                joinColumns = [JoinColumn(name = "codigo_usuario")],
                inverseJoinColumns = [JoinColumn(name = "codigo_papel")]
        )
        @field:Size(min = 1, message = "O usuário deve ter ao menos um papel no sistema")
        var papeis: MutableList<Papel> = mutableListOf()
) : Serializable {

    fun adicionarPapel(papel: Papel) {
        papeis.add(papel)
    }

    fun removerPapel(papel: Papel) {
        papeis.remove(papel)
    }

    override fun hashCode(): Int = Objects.hash(codigo)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false
        val that = other as Usuario
        return codigo == that.codigo
    }

    override fun toString(): String {
        return "codigo: $codigo\nnome: $nome\nemail: $email\nsenha: $senha\nusuario: $nomeUsuario\ndataNascimento: $dataNascimento\nativo: $ativo"
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
